source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,330 @@
+
+// String type requires byte with zero value to indicate end of string.
+class String {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
+    auto index = data.search(&zero, 1, pos);
+    if (index == -1) {
+      return false;
+    }
+
+    // Reserve that many bytes in the string.
+    auto size = index - pos;
+    value_.resize(size);
+    // Now copy from buffer to string.
+    data.copyOut(pos, index - pos, value_.data());
+    pos += (size + 1);
+    left -= (size + 1);
+
+    return true;
+  }
+
+  std::string toString() const { return absl::StrCat("[", value_, "]"); }
+
+private:
+  std::string value_;
+};
+
+// ByteN type is used as the last type in the Postgres message and contains
+// sequence of bytes. The length must be deduced from message length.
+class ByteN {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if (left > (data.length() - pos)) {
+      return false;
+    }
+    value_.resize(left);
+    data.copyOut(pos, left, value_.data());
+    pos += left;
+    left = 0;
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out = "[";
+    out.reserve(value_.size() * 3);
+    bool first = true;
+    for (const auto& i : value_) {
+      char buf[16];
+      if (first) {
+        sprintf(buf, "%02d", i);
+        first = false;
+      } else {
+        sprintf(buf, " %02d", i);
+      }
+      absl::StrAppend(&out, buf);
+    }
+    absl::StrAppend(&out, "]");
+    return out;
+  }
+
+private:
+  std::vector<uint8_t> value_;
+};
+
+// Class represents the structure consisting of 4 bytes of length
+// indicating how many bytes follow.
+// In Postgres documentation it is described as:
+// Int32
+// The length of the function result value, in bytes (this count does not include itself). Can be
+// zero. As a special case, -1 indicates a NULL function result. No value bytes follow in the NULL
+// case.
+//
+// ByteN
+// The value of the function result, in the format indicated by the associated format code. n is the
+// above length.
+class VarByteN {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((left < sizeof(int32_t)) || ((data.length() - pos) < sizeof(int32_t))) {
+      return false;
+    }
+    len_ = data.peekBEInt<int32_t>(pos);
+    pos += sizeof(int32_t);
+    left -= sizeof(int32_t);
+    if (len_ < 1) {
+      // Nothing follows.
+      value_.clear();
+      return true;
+    }
+    if ((left < static_cast<uint64_t>(len_)) ||
+        ((data.length() - pos) < static_cast<uint64_t>(len_))) {
+      return false;
+    }
+    value_.resize(len_);
+    data.copyOut(pos, len_, value_.data());
+    pos += len_;
+    left -= len_;
+    return true;
+  }
+
+  std::string toString() const {
+    char buf[16];
+    sprintf(buf, "[(%d bytes):", len_);
```

Comment:
This is a matter of preference I believe we can use `fmt::sprintf` here, the same as other places.

Reply:
Thanks for suggesting this. I changed all instances of `sprintf` to `fmt::format` and it drastically simplified the code.